lib/ansible/modules/crypto/openssl_cert.py

```diff
@@ -0,0 +1,996 @@
+            if set(san_dns) != set(self.module.params['subjectAltName_only']):
+                self.module.fail_json(
+                    msg="Not all required entries could be found in subjectAltName (was looking for %s in %s)" %
+                    (self.module.params['subjectAltName_only'], san_dns))
+
+
+class SelfSignedCertificate(Certificate):
+    '''Generate the self-signed certificate.'''
+
+    def __init__(self, module):
+        Certificate.__init__(self, module)
+        self.serial_number = randint(1000, 99999)
+        self.notBefore = module.params['selfsigned_notBefore']
+        self.notAfter = module.params['selfsigned_notAfter']
+        self.selfsigned_digest = module.params['selfsigned_digest']
+        self.backup = module.params['backup']
+        self.request = self.load_csr()
+        self.privatekey = self.load_privatekey()
+        self.certificate = None
+
+    def load_privatekey(self):
+        '''Load the privatekey object from buffer.'''
+
+        privatekey_content = open(self.privatekey_path).read()
+        return crypto.load_privatekey(crypto.FILETYPE_PEM, privatekey_content)
+
+    def load_csr(self):
+        '''Load the certificate signing request object from buffer.'''
+
+        csr_content = open(self.csr_path).read()
+        return crypto.load_certificate_request(crypto.FILETYPE_PEM, csr_content)
+
+    def generate(self):
+
+        if not self.force and os.path.exists(self.dest):
+            self.changed = False
+            return
+
+        cert = crypto.X509()
+        cert.set_serial_number(self.serial_number)
+        cert.gmtime_adj_notBefore(self.notBefore)
+        cert.gmtime_adj_notAfter(self.notAfter)
+        cert.set_subject(self.request.get_subject())
+        cert.set_version(self.request.get_version() - 1)
+        cert.set_pubkey(self.request.get_pubkey())
+        try:
+            # NOTE: This is only available starting from pyOpenSSL >= 0.15
+            cert.add_extensions(self.request.get_extensions())
+        except:
+            e = get_exception()
+            raise CertificateError(e)
+        cert.sign(self.privatekey, self.selfsigned_digest)
+        self.certificate = cert
+
+        backupdest = ""
+        if self.changed and not self.module.check_mode:
+            if self.backup and os.path.exists(self.dest):
+                backupdest = module.backup_local(self.dest)
+            cert_file = open(self.dest, 'w')
+            cert_file.write(crypto.dump_certificate(crypto.FILETYPE_PEM, self.certificate))
+            cert_file.close()
+
+    def dump(self):
+
+        result = {
+            'changed': self.changed,
+            'cert_path': self.dest,
+            'notBefore': self.notBefore,
+            'notAfter': self.notAfter,
+            'serial_number': self.serial_number,
+            'backup': backupdest,
+            'warnings': self.warnings,
+        }
+
+        return result
+
+
+class AssertOnlyCertificate(Certificate):
```

Comment:
provider: assertonly isn't idempotent. It returns change = True when nothing actually changed.